src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -378,69 +365,131 @@ and gen_from_arrow ~loc ~env left right =
         let res, xs = aux xs in
         let obs = observable x in
         (res, [%expr [%e obs] @-> [%e xs]])
-    | x -> (gen_from_type ~loc ~env x, [%expr o_nil])
+    | x -> (gen_from_type ~loc  x, [%expr o_nil])
   in
   let x, obs = aux right in
+  (* TODO: export this in qcheck_generators for https://github.com/c-cube/qcheck/issues/190 *)
   let arb = [%expr QCheck.make [%e x]] in
   [%expr
     QCheck.fun_nary QCheck.Tuple.([%e observable left] @-> [%e obs]) [%e arb]
     |> QCheck.gen]
 
-let gen_from_kind_variant ~loc ~env typ_name xs =
+let gen_from_kind_variant ~loc xs =
```

Comment:
This is a pretty short function with a confusing name (there's both `gen_from_kind_invariant` and `gen_from_invariant` - and it has nothing to do with "kinds" as in type theory).
I think it would be clearer if this was just inlined in `gen_from_type_declaration` (where the other cases also take a couple of lines).

Reply:
`kind_variant` and `variant` refers to nodes in the OCaml Parsetree.
![image](https://user-images.githubusercontent.com/32700018/145539117-494f3574-4616-4ca4-b67a-29e375f321ae.png)
![image](https://user-images.githubusercontent.com/32700018/145539241-e529b16d-187b-406f-85f0-731f2043a047.png)

But yes, the function did become relatively small, I can inline it, done in 414264448a0a0f79980e02dae760690d5dd4aed5.

Reply:
Sorry, my bad! 😬